sw/applications/example_iffifo/main.c

```diff
@@ -0,0 +1,179 @@
+{
+  mmio_region_t iffifo_base_addr = mmio_region_from_addr((uintptr_t)IFFIFO_START_ADDRESS);
+  mmio_region_write32(iffifo_base_addr, IFFIFO_INTERRUPTS_REG_OFFSET, 0b0);
+  iffifo_intr_flag = 1;
+  PRINTF(" ** REACH intr. fired.\n");
+}
+
+static dma_target_t tgt_src;
+static dma_target_t tgt_dst;
+static dma_trans_t trans;
+
+int compare_print_fifo_array(void) {
+  int errors = 0;
+  PRINTF("from_fifo = {");
+  for (int i = 0; i < 4; i+=1) {
+    PRINTF("%d",from_fifo[i]);
+    if(i != 4-1) {PRINTF(", ");};
+    if (to_fifo[i]+1 != from_fifo[i]) {++errors;}
+  }
+  PRINTF("}\n");
+  return errors;
+}
+
+void print_status_register(void)
+{
+  mmio_region_t iffifo_base_addr = mmio_region_from_addr((uintptr_t)IFFIFO_START_ADDRESS);
+  int32_t status = mmio_region_read32(iffifo_base_addr, IFFIFO_STATUS_REG_OFFSET);
+  PRINTF("STATUS = ");
+  PRINTF(status & 1 ? "E" : "-"); // FIFO empty
+  PRINTF(status & 2 ? "A" : "-"); // Data available in FIFO
+  PRINTF(status & 4 ? "R" : "-"); // Watermark reached
+  PRINTF(status & 8 ? "F" : "-"); // FIFO full
+  PRINTF("\n");
+}
+
+int main(int argc, char *argv[]) {
+
+    mmio_region_t iffifo_base_addr = mmio_region_from_addr((uintptr_t)IFFIFO_START_ADDRESS);
+    
+    // Enable interrupt on processor side
+    // Enable global interrupt for machine-level interrupts
+    CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+    // Set mie.MEIE bit to one to enable machine-level external interrupts
+    const uint32_t mask = 1 << 11;
+    CSR_SET_BITS(CSR_REG_MIE, mask);
+    
+    if(plic_Init()) {return EXIT_FAILURE;};
+    if(plic_irq_set_priority(EXT_INTR_1, 1)) {return EXIT_FAILURE;};
+    if(plic_irq_set_enabled(EXT_INTR_1, kPlicToggleEnabled)) {return EXIT_FAILURE;};
+    
+    mmio_region_write32(iffifo_base_addr, IFFIFO_WATERMARK_REG_OFFSET, 2);
+    mmio_region_write32(iffifo_base_addr, IFFIFO_INTERRUPTS_REG_OFFSET, 0b1);
+    
+    dma_config_flags_t ret;
+
+     // -- DMA CONFIGURATION --
+
+    dma_init(NULL);
+    tgt_src.ptr        = to_fifo;
+    tgt_src.inc_du     = 1;
+    tgt_src.trig       = DMA_TRIG_MEMORY;
+    tgt_src.type       = DMA_DATA_TYPE_WORD;
+    tgt_src.size_du    = 6;
+
+    tgt_dst.ptr        = IFFIFO_START_ADDRESS + IFFIFO_FIFO_IN_REG_OFFSET;
+    tgt_dst.inc_du     = 0;
+    tgt_dst.trig       = DMA_TRIG_SLOT_EXT_TX;
+    tgt_dst.type       = DMA_DATA_TYPE_WORD;
+    
+    trans.src        = &tgt_src;
+    trans.dst        = &tgt_dst;
+    trans.end        = DMA_TRANS_END_INTR;
+
+    ret = dma_validate_transaction( &trans, DMA_ENABLE_REALIGN, DMA_PERFORM_CHECKS_INTEGRITY );
+    if (ret != 0) {return EXIT_FAILURE;}
+    ret = dma_load_transaction(&trans);
+    if (ret != 0) {return EXIT_FAILURE;}
+    
+    if (compare_print_fifo_array() != 4) {return EXIT_FAILURE;}
+    
+    print_status_register();
+    
+    PRINTF("Launch MM -> Stream DMA\n");
+    dma_launch( &trans );
+    
+    int32_t read0 = mmio_region_read32(iffifo_base_addr, IFFIFO_FIFO_OUT_REG_OFFSET);
```

Comment:
if you are making these 2 read operations to empty the FIFO (pop operation) to let the DMA terminate, please add it in a comment

Reply:
before reading the fifo, can you do "while(fifo_full == 0);" so that you read only if the FIFO is full? of course you have to do it twice

Reply:
actually you do not need that as the reading from the FIFO should be blocking in case it's empty, right?

Reply:
as the fifo is not blocking, reading to an empty fifo succeeds, so the requirements about reading if fiflo_full is strong, but could also be "if not empty"

Reply:
Missed that, yes, totally makes sense! I'll put `while(fifo_full == 0);` as I want the DMA to be blocked.